### PR TITLE
BUG: setitem with boolean mask and tz-aware DTI

### LINKED ISSUE #16889: Boolean masks on dataframes with indexes with timezones introduces NaNs
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np
mask = np.array([False, True, True, False])

idx = pd.date_range('20010101', '20020101')[:4].tz_localize('UTC')
df = pd.DataFrame({'a' : np.arange(4)}, index=idx).astype('float64')
fill = 2 * df
df.loc[mask, :] = fill.loc[mask, :]
print df

idx = pd.date_range('20010101', '20020101')[:4]
df = pd.DataFrame({'a' : np.arange(4)}, index=idx).astype('float64')
fill = 2 * df
df.loc[mask, :] = fill.loc[mask, :]
print df
```
#### Problem description

This is the output that I get:

```
                            a
2001-01-01 00:00:00+00:00   0
2001-01-02 00:00:00+00:00 NaN
2001-01-03 00:00:00+00:00 NaN
2001-01-04 00:00:00+00:00   3
            a
2001-01-01  0
2001-01-02  2
2001-01-03  4
2001-01-04  3

```

In the first case, a timezone of UTC seems to cause the masking operation to introduce NaNs.  The second case is correct.  This problem seems to be introduced in pandas 0.17.1, whereas with 0.16 I get the following output below

#### Expected Output

```
                           a
2001-01-01 00:00:00+00:00  0
2001-01-02 00:00:00+00:00  2
2001-01-03 00:00:00+00:00  4
2001-01-04 00:00:00+00:00  3
            a
2001-01-01  0
2001-01-02  2
2001-01-03  4
2001-01-04  3
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Darwin
OS-release: 14.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: None
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.10.2
scipy: 0.16.0
statsmodels: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.4.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
Jinja2: None

</details>

### PR DESCRIPTION
**Pull request by hhuuggoo**
closes #16889 

 - [x] closes #xxxx
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexing/test_datetime.py [hunk 1]
```diff
@@ -8,6 +8,14 @@
 
 class TestDatetimeIndex(object):
 
+    def test_setitem_with_datetime_tz(self):
+        mask = np.array([True, False, True, False])
```
Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: add a comment with the issue number

can you also add the working example as well (from your issue)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. pls add a whatnew note in indexing bug fixes.

File: pandas/tests/indexing/test_datetime.py (hunk 1 above)
Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: as well as the case of 

``testf.loc[mask] = df.loc[mask]``

File: pandas/tests/indexing/test_datetime.py [hunk 2]
```diff
@@ -8,6 +8,14 @@
...
+        idx = pd.date_range('20010101', periods=4, tz='UTC')
+        df = pd.DataFrame({'a': np.arange(4)}, index=idx).astype('float64')
+        test = df.copy()
```
Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: call this ``result`` rather than ``test``

Comment (hhuuggoo) [inline pandas/tests/indexing/test_datetime.py]: will do - thanks

File: pandas/tests/indexing/test_datetime.py [hunk 3]
```diff
@@ -8,6 +8,40 @@
 
 class TestDatetimeIndex(object):
 
+    def test_tuple_setitem_with_datetime_tz(self):
```
Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: sorry I meant put these cases in a single test (obviously with several sub-sections, and avoid duplicated code within reason)

Comment (hhuuggoo) [inline pandas/tests/indexing/test_datetime.py]: Oh yea no problem

Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: great

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. just need a whatsnew note. ping on green. (but fix section, indexing is fine) for 0.21.0

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -144,6 +144,7 @@ Bug Fixes
 ~~~~~~~~~
 
 - Fixes regression in 0.20, :func:`Series.aggregate` and :func:`DataFrame.aggregate` allow dictionaries as return values again (:issue:`16741`)
+- Fixes `__setitem__` with a tuple for dataframes with a tz-aware `DatetimeIndex` (:issue: `16889`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: this is not correct:

``DataFrame.loc`` for setting wth alignment and a tz-aware DatetimeIndex (use double-back ticks around DTI)

Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: no space between `` :issue:`16889` ``

File: pandas/tests/indexing/test_datetime.py [hunk 5]
```diff
@@ -8,6 +8,33 @@
 
 class TestDatetimeIndex(object):
 
+    def test_setitem_with_datetime_tz(self):
+        # 16889
+        # support setitem with datetimeindex with tz
```
Comment (jreback) [inline pandas/tests/indexing/test_datetime.py]: ``.loc`` with alignment

**Review by jreback — CHANGES_REQUESTED** [review]
doc changes

**Comment by jreback** [comment]
can you rebase and update

**Comment by hhuuggoo** [comment]
will do later today

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**